### DISCUSSION TITLE
Building UI with YAML

### DISCUSSION THREAD

**Discussion by darcyg** [flet-dev/flet#441] [Ideas]
Similar to QT's *.UI file, build the UI of the application from Yaml
There is only logic code inside Python.
Of course it's better to provide command line tools that generate python UI config via yaml.
In addition, it is recommended to use nuitke packaging.

**Comment by FeodorFitsner** [comment]
I thought about that, like describing UI in YAML. But is it really going to be less verbose than a plain Python. With any DSL you loose syntax checking, completion, some math in prop values, etc. no?

**Reply by darcyg** [reply]
I think it is better to load the UI layout through a special UI file.
It is also suitable for a graphical editor UI in the future, just like REMI.
I used REMI's WYSIWYG to do WebUI before. Every time after editing the layout, the generated UI code and the python file I wrote have tons of conflicts. Manual readjustment is required.
So I think it is like most desktop graphical IDE development tools, such as VB, DELPHI, QT
It would be better to separate UI layout files from program logic, especially when complex UI logic is modified.
When you have dozens or hundreds of controls on a UI layout, a separate UI configuration file is a better choice.

**Comment by ghost** [comment]
IMHO, if a sane UI DSL is desirable, it first need a good native implementation, (in Python, without YAML, XML, JSON, etc). I have some experience with a Web framework (Plone/Zope) that defines lots of things in Python, XML, templates, CSS, etc, and it can be quite difficult to manage large projects with all the different modules. Python can be used to maintain compact definitions, so I tend to resist the temptation of using available DSLs (unless required); extra levels of abstractions are often more difficult to understand and maintain in the long term, and require more code. Of course it depends on what's being abstracted, but for a young and ambitious project like flet I believe that Python is more than enough.

**Reply by darcyg** [reply]
I understand your thinking.
I found and used flet because it doesn't require me to learn various front-end frameworks and can handle interactive WEBUI with logic I'm familiar with. Some of my work can be carried out without relying on front-end engineers. If it is just a simple UI interface, it can be done well with python. But if it is a complex UI interface, especially considering the cross-platform and cross-language. Let the UI layout be loaded from the configuration file. would make the problem easier.
The direction of flet is a trend, and it is indeed sufficient for small projects now. But I think if flet is to be used by more people, especially in certain fields, most of the users are hobbyists, and a WYSIWYG UI editor will inevitably be needed. At that time, separation of UI layout and business logic was definitely a good way. If you think the trend is inevitable, it is better to add these from the beginning, so as not to change the API too much when it is added in the future.
Personally, I have been looking for a modern WEB tool. I often need to add some WEB-based tools to embedded systems (such as iot edge computing, routing) to facilitate users to call some underlying controls. I have used luci such as openwrt. But luci itself is designed for routing, and the default functions are too many, too complicated, and difficult to tailor. Before getting to know pglet/flet, I also used remi. His ui is too traditional and not suitable for mobile phone control.

**Comment by ghost** [comment]
I will reformulate my previous comment: a working WYSIWYG UI editor could be something good *only* if it's not required, and if it can be added after (or in parallel), always as an option. I understand the "user" paradigm, and it can be prioritized in certain contexts, but WYSIWYG does not mean "modern", and it certainly mean "more complex". Also, think about the difference between a parametric and a modeling CAD; playdoh is fun, but it doesn't last. Maybe I'm too old to trust WYSIWYG, previously traumatized by horrible and pretentious software like DreamWeaver. And I see there's more! https://en.wikipedia.org/wiki/List_of_HTML_editors#WYSIWYG_editors
